src/cargo/ops/cargo_rustc/context.rs

```diff
@@ -358,9 +366,18 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
                 })
             } else {
                 None
+                /*
```

Comment:
Hm yes I can see how this would indeed be a problem! For executables we could generate ones with long filenames and them link them into a location without the metadata, but that does weird things with dynamic libraries unfortunately, so we can't do that everywhere.

Short of building everything in an entirely new directory, we may have to just stomach recompiling these top level dependencies whenever various flags change.